docs/ORTModule_ModuleWithLoss_Wrapper.md

```diff
@@ -0,0 +1,70 @@
+# ONNX Runtime ModuleWithLoss Wrapper
+
+This document provides instructions on implementing a wrapper similar to the ModuleWithLoss Wrapper in Optimum. By implementing this wrapper, you can compute the loss inside ONNX Runtime (ORT), enabling you to leverage additional optimizations such as label sparsity optimization.
+## Implementation Steps
+
+Follow these steps to create your own ModuleWithLoss for computing loss inside ONNX Runtime:
+
+### Step 1: Define `ModuleWithLoss` Class
+
+1. Create a class named `ModuleWithLoss` that extends `nn.Module`.
+2. Implement the `__init__` method to initialize the wrapper with the original model.
+3. Implement the `forward` method to perform the forward pass of the model and return the outputs.
+
+### Step 2: Compute Loss Function
+
+1. Implement a function named `compute_loss` that takes the model, inputs, and targets as arguments.
+2. Inside the function, compute the forward pass of the model using the inputs.
+3. Compute the loss between the model outputs and the targets.
+4. Return the computed loss.
+
+### Step 3: Training Loop
+
+1. Create an instance of `ModuleWithLoss` by passing the original model as an argument.
+2. Initialize the optimizer for training.
+3. Enter the training loop and iterate over the training data.
+4. Zero the gradients of the optimizer.
+5. Compute the loss by calling the `compute_loss` function and passing the model, inputs, and targets.
+6. Perform the backward pass and optimization step by calling `loss.backward()` and `optimizer.step()`.
+
+
+## Full Example
+
```

Comment:
```diff

    # Define the model architecture
    class PretrainedModel(nn.Module):
        ...

        def forward(self, input_ids, attention_mask):
            ...
            transformer_outputs = self.transformer(
                input_ids,
                attention_mask=attention_mask,
                ...
            )
            hidden_states = transformer_outputs[0]
            lm_logits = self.lm_head(hidden_states)
            return lm_logits

+    class ModuleWithLoss(nn.Module):
+        def __init__(self, model):
+            super().__init__()
+            self.model = model

+        def forward(self, *inputs):
+            # Perform the forward pass of the model and return the outputs
+            outputs = self.model(inputs)

+            # Compute the forward pass of the model
+            lm_logits= model(inputs)

+            # Compute the cross-entropy loss
+            loss = nn.CrossEntropyLoss()(lm_logits, labels)
+            return loss


    # Training loop
    model = PretrainedModel(...)
+    model = ModuleWithLoss(model)

    optimizer = torch.optim.Adam(model.parameters(), lr=0.001)

    model = ORTModule(model)

    for inputs, labels in dataloader:
        optimizer.zero_grad()

-        # Compute the forward pass of the model
-        lm_logits= model(inputs)

-        # Compute the cross-entropy loss
-        loss = nn.CrossEntropyLoss()(outputs, labels)


+        # Compute the forward pass and cross-entropy loss
+        loss= model(inputs)

        # Backward pass and optimization step
        loss.backward()
        optimizer.step()

```

Reply:
I feel the example script is not explaining loss wrapping correctly. I put an example above, which is a diff view, comparing origional model train script with the modified model train script. Does that looks better to be self-explained?

Reply:
The steps need to be adapted according to it.

Reply:
Thanks for the suggestion, will modify it accordingly.
Thanks